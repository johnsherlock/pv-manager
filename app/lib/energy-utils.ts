import { convertJoulesToKwh } from './num-utils';
import { MinutePVData, HourlyPVData, HalfHourlyPVData } from '../model/pv-data';

export const calculateEnergyConsumption = (
  importedJoules: number = 0,
  generatedJoules: number = 0,
  immersionDivertedJoules: number = 0,
  exportedJoules: number = 0): number => {

  return importedJoules + generatedJoules - immersionDivertedJoules - exportedJoules;
};

export const calculateGreenEnergyPercentage = (importedEnergy: number = 0, consumedEnergy: number = 0): number => {
  if (importedEnergy === 0) {
    return 100;
  }
  const greenEnergy = consumedEnergy - importedEnergy;
  if (greenEnergy <= 0) {
    return 0;
  }
  const percentage = Math.round((greenEnergy / consumedEnergy) * 100);
  return percentage;
};

export const convertMinuteDataToHourlyData = (minuteData: MinutePVData[] = []): HourlyPVData[] => {
  const hourlyTotals: { [hour: number]: HourlyPVData } = {};

  for (const minuteItem of minuteData) {
    const {
      hour, importedJoules: importJoules, generatedJoules, exportedJoules: exportJoules,
      immersionDivertedJoules: immersionDivertedJoules, immersionBoostedJoules: immersionBoostedJoules, consumedJoules,
    } = minuteItem;

    hourlyTotals[hour] ??= {
      ...minuteItem,
      minute: 0,
      importedKwH: 0,
      generatedKwH: 0,
      exportedKwH: 0,
      immersionDivertedKwH: 0,
      immersionBoostedKwH: 0,
      immersionDivertedMins: 0,
      immersionBoostedMins: 0,
      consumedKwH: 0,
    };
    hourlyTotals[hour].importedKwH += convertJoulesToKwh(importJoules);
    hourlyTotals[hour].generatedKwH += convertJoulesToKwh(generatedJoules);
    hourlyTotals[hour].exportedKwH += convertJoulesToKwh(exportJoules);
    hourlyTotals[hour].immersionDivertedKwH += convertJoulesToKwh(immersionDivertedJoules);
    hourlyTotals[hour].immersionBoostedKwH += convertJoulesToKwh(immersionBoostedJoules);
    hourlyTotals[hour].immersionDivertedMins += immersionDivertedJoules > 0 ? 1 : 0;
    hourlyTotals[hour].immersionBoostedMins += immersionBoostedJoules > 0 ? 1 : 0;
    hourlyTotals[hour].consumedKwH += convertJoulesToKwh(consumedJoules);
  }
  // calculate the green energy percentage for each half hour
  for (const halfHour in hourlyTotals) {
    const { importedKwH: importKwH, consumedKwH } = hourlyTotals[halfHour];
    hourlyTotals[halfHour].greenEnergyPercentage = calculateGreenEnergyPercentage(importKwH, consumedKwH);
  }
  return Object.values(hourlyTotals);
};

export const convertMinuteDataToHalfHourlyData = (minuteData: MinutePVData[] = []): HalfHourlyPVData[] => {
  const halfHourlyTotals: { [halfHour: string]: HalfHourlyPVData } = {};

  for (const minuteItem of minuteData) {
    const {
      hour, minute, importedJoules: importJoules, generatedJoules, exportedJoules: exportJoules,
      immersionDivertedJoules: immersionDivertedJoules, immersionBoostedJoules: immersionBoostedJoules, consumedJoules,
    } = minuteItem;
    const halfHour = `${hour}:${minute < 30 ? '00' : '30'}`;
    halfHourlyTotals[halfHour] ??= {
      ...minuteItem,
      hour: parseInt(halfHour.split(':')[0]),
      minute: parseInt(halfHour.split(':')[1]) == 0 ? 0 : 30,
      importedKwH: 0,
      generatedKwH: 0,
      exportedKwH: 0,
      immersionDivertedKwH: 0,
      immersionBoostedKwH: 0,
      consumedKwH: 0,
      immersionDivertedMins: 0,
      immersionBoostedMins: 0,
      greenEnergyPercentage: 0,
    };
    halfHourlyTotals[halfHour].importedKwH += convertJoulesToKwh(importJoules);
    halfHourlyTotals[halfHour].generatedKwH += convertJoulesToKwh(generatedJoules);
    halfHourlyTotals[halfHour].exportedKwH += convertJoulesToKwh(exportJoules);
    halfHourlyTotals[halfHour].immersionDivertedKwH += convertJoulesToKwh(immersionDivertedJoules);
    halfHourlyTotals[halfHour].immersionBoostedKwH += convertJoulesToKwh(immersionBoostedJoules);
    halfHourlyTotals[halfHour].immersionDivertedMins += immersionDivertedJoules > 0 ? 1 : 0;
    halfHourlyTotals[halfHour].immersionBoostedMins += immersionBoostedJoules > 0 ? 1 : 0;
    halfHourlyTotals[halfHour].consumedKwH += convertJoulesToKwh(consumedJoules);
    halfHourlyTotals[halfHour].greenEnergyPercentage += calculateGreenEnergyPercentage(importJoules, consumedJoules);
  }
  // calculate the green energy percentage for each half hour
  for (const halfHour in halfHourlyTotals) {
    const { importedKwH: imp, consumedKwH: conp } = halfHourlyTotals[halfHour];
    halfHourlyTotals[halfHour].greenEnergyPercentage = calculateGreenEnergyPercentage(imp, conp);
  }
  return Object.values(halfHourlyTotals);
};