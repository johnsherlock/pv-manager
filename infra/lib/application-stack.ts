import * as amplify from '@aws-cdk/aws-amplify-alpha';
import * as cdk from 'aws-cdk-lib';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import { Construct } from 'constructs';

export class ApplicationStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const amplifyApp = new amplify.App(this, 'pv-manager ', {
      sourceCodeProvider: new amplify.GitHubSourceCodeProvider({
        owner: 'johnsherlock',
        repository: 'pv-manager',
        oauthToken: cdk.SecretValue.secretsManager('pvm', {
          jsonField: 'github-token',
        }),
      }),
      buildSpec: codebuild.BuildSpec.fromObjectToYaml({
        version: '1.0',
        frontend: {
          phases: {
            preBuild: {
              commands: [
                'npm ci',
              ],
            },
            build: {
              commands: [
                'npm run build',
              ],
            },
          },
          artifacts: {
            baseDirectory: 'src',
            files:
              - '**/*',
          },
          cache: {
            paths:
              - 'node_modules/**/*',
          },
        },
      }),
    });
    const mainBranch = amplifyApp.addBranch('main');
  }
}
