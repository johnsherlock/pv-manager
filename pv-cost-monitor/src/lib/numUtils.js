export const formatDecimal = number => Math.round(number * 100) / 100;
export const convertJoulesToKwh = joules => joules ? (joules / 3600000) : '';