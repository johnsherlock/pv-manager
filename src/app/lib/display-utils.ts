export const isTouchScreen = () => 'ontouchstart' in window || navigator.maxTouchPoints > 0;