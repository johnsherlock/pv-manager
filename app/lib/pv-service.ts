import axios from 'axios';
import { EddiData } from '../../shared/eddi-data';
import { mapEddiDataToMinutePVData } from '../../shared/energy-utils';
import { MinutePVData } from '../../shared/pv-data';

export interface DailyEnergyUsageProps {
  data: MinutePVData[];
}

export const getPVDataForDate = async (formattedTargetDate: string): Promise<MinutePVData[]> => {
  console.log(`Retrieving per minute data for ${formattedTargetDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/minute-data?date=${formattedTargetDate}`;
  // const url = `http://localhost:3001/minute-data?date=${formattedTargetDate}`;
  try {
    const response = await axios.get(url);
    const data = await Promise.resolve(response.data as EddiData[]);
    return data.map(mapEddiDataToMinutePVData);
  } catch (error) {
    console.log('Error retrieving remote data', error);
    document.body.style.cursor = 'auto';
    return [];
  };
};