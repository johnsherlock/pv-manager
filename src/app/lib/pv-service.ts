import axios from 'axios';
import { EddiData } from '../../shared/eddi-data';
import { mapEddiDataToMinutePVData } from '../../shared/energy-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';

export interface DailyEnergyUsageProps {
  data: MinutePVData[];
}

export const getPVDataForDate = async (formattedTargetDate: string): Promise<MinutePVData[]> => {
  console.log(`Retrieving per minute data for ${formattedTargetDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/minute-data?date=${formattedTargetDate}`;
  try {
    const response = await axios.get(url);
    const data = await Promise.resolve(response.data as EddiData[]);
    return data.map(mapEddiDataToMinutePVData);
  } catch (error) {
    console.log(`Error retrieving data for ${formattedTargetDate}`, error);
    document.body.style.cursor = 'auto';
    return [];
  };
};

export const getPVTotalsForRange = async (startDate: string, endDate: string): Promise<Totals[]> => {
  console.log(`Retrieving PV summary data for range ${startDate} - ${endDate}`);
  const url = `https://jmcjm1731b.execute-api.eu-west-1.amazonaws.com/prod/aggregate-historical-data?serialNumber=21494842&startDate=${startDate}&endDate=${endDate}`;
  try {
    const response = await axios.get(url);
    return await Promise.resolve(response.data as Totals[]);
  } catch (error) {
    console.log('Error retrieving totals for range', error);
    document.body.style.cursor = 'auto';
    return [];
  };
};