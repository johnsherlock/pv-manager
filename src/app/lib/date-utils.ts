import moment from 'moment';
import { CalendarScale, DateRange, FormattedDateRange } from './state-utils';
import { BasePVData } from '../../shared/pv-data';

export const formatDate = (date: moment.Moment): string => moment(date).startOf('day').format('YYYY-MM-DD');

export const dawnOfTime = moment(new Date('2023-01-20')).startOf('day');

export const toMoment = (pvData: BasePVData, locale?: string): moment.Moment => {
  return moment({
    year: pvData.year,
    month: pvData.month - 1,
    date: pvData.dayOfMonth,
    hour: pvData.hour,
    minute: pvData.minute,
  });
};

export const getFormattedTime = (pvData: BasePVData, locale?: string): string => {
  const time = toMoment(pvData, locale);
  return time.format('HH:mm');
};

export const getDateRange = (scale: CalendarScale, date: Date): DateRange => {

  switch (scale) {
    case 'week':
    case 'month':
    case 'year': {
      const m = moment(date);
      const start = moment(m.startOf(scale).toDate());
      const end = moment(m.endOf(scale).toDate());
      return { startDate: start, endDate: end };
    }
    default:
      return { startDate: moment(date), endDate: (moment(date)) };
  }
};

export const formatDateRange = (dateRange: DateRange): FormattedDateRange => {
  return {
    startDate: formatDate(dateRange.startDate),
    endDate: formatDate(dateRange.endDate),
  };
};