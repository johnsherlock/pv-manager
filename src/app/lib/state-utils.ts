import moment from 'moment';
import * as dateUtils from './date-utils';
import { MinutePVData, Totals } from '../../shared/pv-data';
import { Scale, View } from '../energy-usage-line-graph';

export type CalendarScale = 'day' | 'week' | 'month' | 'year' | 'custom';

export interface AppState {
  intervalId?: number | null;
  today: moment.Moment;
  selectedDate: moment.Moment;
  startDate?: moment.Moment;
  endDate?: moment.Moment;
  formattedSelectedDate: string;
  pvDataCache: Map<string, MinutePVData[]>;
  totals: Map<string, Totals>;
  energyUsageLineGraphScale: Scale;
  energyUsageLineGraphView: View;
  calendarScale: CalendarScale;
}

export const initialState = (): AppState => {
  const today: moment.Moment = moment().startOf('day');
  return {
    today: today,
    selectedDate: today,
    formattedSelectedDate: dateUtils.formatDate(today),
    pvDataCache: new Map(),
    totals: new Map(),
    energyUsageLineGraphScale: 'hour',
    energyUsageLineGraphView: 'line',
    calendarScale: 'day',
  };
};