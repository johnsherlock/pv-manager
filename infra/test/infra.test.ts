import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ApplicationStack } from '../lib/application-stack';

test('Account Governer Stack matches snapshot', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new ApplicationStack(app, 'MyTestStack');
  // THEN

  const template = Template.fromStack(stack);

  // template should match snapshot
  expect(template).toMatchSnapshot();
});
