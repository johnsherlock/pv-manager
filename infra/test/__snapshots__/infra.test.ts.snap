// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Governer Stack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "pvmanager7AB39DE7": {
      "Properties": {
        "BasicAuthConfig": {
          "EnableBasicAuth": false,
        },
        "BuildSpec": "version: "1.0"
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
    build:
      commands:
        - npm run build
  artifacts:
    baseDirectory: src
    files: .nan
  cache:
    paths: .nan
",
        "IAMServiceRole": {
          "Fn::GetAtt": [
            "pvmanagerRole57D61FC7",
            "Arn",
          ],
        },
        "Name": "pv-manager ",
        "OauthToken": "{{resolve:secretsmanager:pvm:SecretString:github-token::}}",
        "Repository": "https://github.com/johnsherlock/pv-manager",
      },
      "Type": "AWS::Amplify::App",
    },
    "pvmanagerRole57D61FC7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "amplify.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "pvmanagermain16482FDF": {
      "Properties": {
        "AppId": {
          "Fn::GetAtt": [
            "pvmanager7AB39DE7",
            "AppId",
          ],
        },
        "BranchName": "main",
        "EnableAutoBuild": true,
        "EnablePullRequestPreview": true,
      },
      "Type": "AWS::Amplify::Branch",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
